File: dynamic-array/dynamicarray_test.lua
local dynamicArray = require("dynamicarray")

local function assertEquals(expected, va)
	assert(expected == va, "Expected " .. expected .. "but received" .. va)
end

local function testCase1()
	local arr = dynamicArray(1)

	assertEquals(0, arr.getSize())
	assertEquals(1, arr.getCapacity())
end

local function testCase2()
	local arr = dynamicArray(1)

	arr.pushBack(1)
	assertEquals(1, arr.getSize())
	assertEquals(1, arr.getCapacity())

	arr.pushBack(2)
	assertEquals(2, arr.getSize())
	assertEquals(2, arr.getCapacity())
end

local function testCase3()
	local arr = dynamicArray(1)

	arr.pushBack(1)
	arr.pushBack(2)

	assertEquals(2, arr.get(1))

	arr.set(1, 3)
	assertEquals(3, arr.get(1))

	assertEquals(3, arr.popBack())

	assertEquals(1, arr.getSize())
	assertEquals(2, arr.getCapacity())
end

testCase1()
testCase2()
testCase3()

File: dynamic-array/dynamicarray_test.go
package main

import "testing"

func checkSizeAndCapacity(
	t *testing.T,
	arr *DynamicArray,
	expectedSize int,
	expectedCapacity int,
) {
	size := arr.GetSize()
	if size != expectedSize {
		t.Fatalf("Size was %d but %d is expected", size, expectedSize)
	}

	capacity := arr.GetCapacity()
	if capacity != expectedCapacity {
		t.Fatalf("Capacity was %d but %d is expected", capacity, expectedCapacity)
	}

}
func TestCase1(t *testing.T) {
	arr := NewDynamicArray(1)

	checkSizeAndCapacity(t, arr, 0, 1)
}

func TestCase2(t *testing.T) {
	arr := NewDynamicArray(1)

	arr.PushBack(1)
	checkSizeAndCapacity(t, arr, 1, 1)

	arr.PushBack(2)
	checkSizeAndCapacity(t, arr, 2, 2)
}

//"popback", "getSize", "getCapacity"]

func TestCase3(t *testing.T) {
	arr := NewDynamicArray(1)

	arr.PushBack(1)
	arr.PushBack(2)

	value := arr.Get(1)
	if value != 2 {
		t.Fatalf("Value was %d but %d was expected", value, 2)
	}

	arr.Set(1, 3)
	value = arr.Get(1)
	if value != 3 {
		t.Fatalf("Value was %d but %d was expected", value, 3)
	}

	value = arr.PopBack()
	if value != 3 {
		t.Fatalf("Value was %d but %d was expected", value, 3)
	}

	checkSizeAndCapacity(t, arr, 1, 2)
}
